### PR TITLE
Mellow fallback

### PR DESCRIPTION
**Pull request by hweawer**

### CODE REVIEW

File: src/blockchain/deposit_strategy/base_deposit_strategy.py [hunk 1]
```diff
@@ -44,3 +44,17 @@ def _depositable_ether(self) -> Wei:
...
         depositable_ether += additional_ether
         return depositable_ether
+
+    def deposited_keys_amount(self, module_id: int) -> int:
+        depositable_ether = self._depositable_ether()
+        possible_deposits_amount_assumption = self.w3.lido.staking_router.get_staking_module_max_deposits_count(
+            module_id,
+            depositable_ether,
+        )
+        possible_deposited_eth = 32 * possible_deposits_amount_assumption
```
Comment (F4ever) [inline src/blockchain/deposit_strategy/base_deposit_strategy.py:54]: `32 ETH`
`32 * 10**18`
or Web3.to_wei(32, 'eth')

File: src/blockchain/deposit_strategy/deposit_transaction_sender.py [hunk 2]
```diff
@@ -28,50 +42,115 @@ def _prepare_signs_for_deposit(quorum: list[DepositMessage]) -> tuple[tuple[str,
 
     def prepare_and_send(
         self,
+        module_id: int,
```
Comment (F4ever) [inline src/blockchain/deposit_strategy/deposit_transaction_sender.py:45]: Let's use order params like in `prepare_and_send`
Or also change `prepare_and_send` annotation to (int, list)

File: src/blockchain/deposit_strategy/deposit_transaction_sender.py [hunk 3]
```diff
@@ -28,50 +42,115 @@ def _prepare_signs_for_deposit(quorum: list[DepositMessage]) -> tuple[tuple[str,
...
 
-    def _prepare_general_tx(self, quorum: list[DepositMessage]):
+    def _send_transaction(
+        self,
+        tx: ContractFunction,
+        flashbots_works: bool
+    ) -> bool:
+        return self._w3.transaction.check(tx) and self._w3.transaction.send(tx, flashbots_works, self._TIMEOUT_IN_BLOCKS)
+
+    def _sender_checks(self, module_id: int) -> bool:
```
Comment (F4ever) [inline src/blockchain/deposit_strategy/deposit_transaction_sender.py:84]: Not sure `Sender` should know anything about strategies and etc.
Looks like `Sender` should prepare and send transaction

File: src/blockchain/deposit_strategy/deposit_transaction_sender.py [hunk 4]
```diff
@@ -16,9 +23,16 @@ class Sender:
...
     _TIMEOUT_IN_BLOCKS = 6
+    _next_sender: Sender = None
 
-    def __init__(self, w3: Web3):
+    def __init__(self, w3: Web3, gas_price_calculator: GasPriceCalculator, strategy: BaseDepositStrategy):
         self._w3 = w3
+        self._gas_price_calculator = gas_price_calculator
+        self._strategy = strategy
+
+    def add_sender(self, sender: Sender):
```
Comment (F4ever) [inline src/blockchain/deposit_strategy/deposit_transaction_sender.py:33]: Linked list is not obvious (with recursion), better use just list
But I propose we can do it with one if only. Not sure if there will be another type of sender

File: src/blockchain/deposit_strategy/deposit_transaction_sender.py [hunk 5]
```diff
@@ -28,50 +42,115 @@ def _prepare_signs_for_deposit(quorum: list[DepositMessage]) -> tuple[tuple[str,
...
+            if tx and self._send_transaction(tx, with_flashbots):
+                return True
+
+        if self._next_sender is not None:
+            return self._next_sender.prepare_and_send(module_id, quorum, with_flashbots)
+
+        return False
 
-    def _prepare_mellow_tx(self, quorum: list[DepositMessage]) -> ContractFunction:
+    def _prepare_tx(self, quorum: list[DepositMessage]) -> Optional[ContractFunction]:
```
Comment (F4ever) [inline src/blockchain/deposit_strategy/deposit_transaction_sender.py:59]: Not good idea that class can be optional. Looks like we should expect `ContractFunction` or exception here

File: src/bots/depositor.py [hunk 6]
```diff
@@ -182,25 +139,15 @@ def _deposit_to_module(self, module_id: int) -> bool:
         gas_is_ok = self._gas_price_calculator.is_gas_price_ok(module_id)
         logger.info({'msg': 'Calculate gas recommendations.', 'value': gas_is_ok})
 
-        strategy, is_mellow = self._select_strategy(module_id)
-        is_deposit_amount_ok = self._gas_price_calculator.calculate_deposit_recommendation(strategy, module_id)
-        logger.info({'msg': 'Calculations deposit recommendations.', 'value': is_deposit_amount_ok})
-
-        if is_depositable and quorum and can_deposit and gas_is_ok and is_deposit_amount_ok:
-            logger.info({'msg': 'Checks passed. Prepare deposit tx.', 'is_mellow': is_mellow})
-            success = self.prepare_and_send_tx(quorum, is_mellow, self._flashbots_works)
```
Comment (F4ever) [inline src/bots/depositor.py:191]: Maybe better was just adding here smth like
```
success = self.prepare_and_send_tx(quorum, is_mellow, self._flashbots_works)

if not success and is_mellow:
    self.prepare_and_send_tx(quorum, False, self._flashbots_works)
```

And remove `_mellow_works` from class Depositor?

File: src/blockchain/deposit_strategy/deposit_transaction_sender.py (hunk 2 above)
Comment (hweawer) [reply src/blockchain/deposit_strategy/deposit_transaction_sender.py]: Done

File: src/blockchain/deposit_strategy/deposit_transaction_sender.py (hunk 3 above)
Comment (hweawer) [reply src/blockchain/deposit_strategy/deposit_transaction_sender.py]: Removed

File: src/blockchain/deposit_strategy/deposit_transaction_sender.py (hunk 4 above)
Comment (hweawer) [reply src/blockchain/deposit_strategy/deposit_transaction_sender.py]: Removed

File: src/blockchain/deposit_strategy/deposit_transaction_sender.py (hunk 5 above)
Comment (hweawer) [reply src/blockchain/deposit_strategy/deposit_transaction_sender.py]: Removed

File: src/bots/depositor.py (hunk 6 above)
Comment (hweawer) [reply src/bots/depositor.py:191]: Done

**Review by F4ever — APPROVED** [review]

**Review by infloop — APPROVED** [review]

**Pull request merged**